src/lib.rs

```diff
@@ -59,6 +66,49 @@ pub struct Pointer<T> {
     pending: PendingRef,
 }
 
+impl<T> Pointer<T> {
+    /// Creates a new `WeakPointer` to this component.
+    pub fn downgrade(&self) -> WeakPointer<T> {
+        let epoch = self.target.read().unwrap().epoch[self.index];
```

Comment:
I'd like to never ever lock the target of a pointer inside the pointer methods. Imagine locking a storage for writing, then trying to `downgrade` one of its pointers. BAM, deadlock!
Instead, every strong pointer should carry the `epoch` internally, passed at creation, since the one it `target.epoch` is guaranteed to be the same for the lifetime of any such pointer.